# **Calidad de Sueño**

####Flores Vázquez Yiyetzi Fabiola
####Grupo 5AM2
####Analítica y Visualización de datos

Los datos se extrajeron de la encuesta anual realizada por el ENSANUT, se tomo la información correspondiente al año 2016, ya que en dicho año, es el único año en el que se incorporaron datos referentes con referencia al sueño.

In [ ]:
import pandas as pd

# Carga del dataset desde el drive
df = pd.read_csv('/content/drive/MyDrive/sueño.csv')

# Vista previa
print(df.head())
print(df.columns)


                  folio  inte  idinsert  entidad  munici  locali    maq  edad  \
0  010010001209A0020131     1   1076190        1       1       1  MC036    67   
1  010010001209A0020161     1   1076193        1       1       1  MC036    55   
2  010010001209A0080071     2   1075466        1       1       1  MC034    44   
3  010010001209A0080101     1   1075538        1       1       1  MC034    47   
4  010010001209A0120041     1   1076278        1       1       1  MC036    36   

   sexo  intsel  ...  horasueno1  horasueno2  horasueno3  clonazepam  \
0     1       1  ...           0           1           0           1   
1     1       1  ...           1           0           0           0   
2     0       1  ...           0           1           0           0   
3     0       1  ...           0           1           0           0   
4     1       1  ...           1           0           0           0   

   diazepam  alprazolam  sedante  cinturaprom   padormir1 padormir2  
0         

<ipython-input-2-19a96d76d926>:4: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/drive/MyDrive/sueño.csv')


In [ ]:
df_info = df.info()
df_head = df.head()
df_description = df.describe(include='all')

df_head, df_info, df_description

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8649 entries, 0 to 8648
Data columns (total 99 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   folio               8649 non-null   object 
 1   inte                8649 non-null   int64  
 2   idinsert            8649 non-null   int64  
 3   entidad             8649 non-null   int64  
 4   munici              8649 non-null   int64  
 5   locali              8649 non-null   int64  
 6   maq                 8649 non-null   object 
 7   edad                8649 non-null   int64  
 8   sexo                8649 non-null   int64  
 9   intsel              8649 non-null   int64  
 10  intsel2             5131 non-null   float64
 11  su101               8639 non-null   float64
 12  su102               8649 non-null   int64  
 13  su103               8649 non-null   int64  
 14  su104               8649 non-null   int64  
 15  su105               3230 non-null   float64
 16  su106 

(                  folio  inte  idinsert  entidad  munici  locali    maq  edad  \
 0  010010001209A0020131     1   1076190        1       1       1  MC036    67   
 1  010010001209A0020161     1   1076193        1       1       1  MC036    55   
 2  010010001209A0080071     2   1075466        1       1       1  MC034    44   
 3  010010001209A0080101     1   1075538        1       1       1  MC034    47   
 4  010010001209A0120041     1   1076278        1       1       1  MC036    36   
 
    sexo  intsel  ...  horasueno1  horasueno2  horasueno3  clonazepam  \
 0     1       1  ...           0           1           0           1   
 1     1       1  ...           1           0           0           0   
 2     0       1  ...           0           1           0           0   
 3     0       1  ...           0           1           0           0   
 4     1       1  ...           1           0           0           0   
 
    diazepam  alprazolam  sedante  cinturaprom   padormir1 padormi

Se llevara a cabo el proceso de transformación del dataset

In [ ]:
#Revisión de porcentajes de valores nulos en nuestras variables

pd.set_option('display.max_rows', None)

valores_nulos = df.isnull().sum() #Suma de valores nulos
porcentaje_nulos = (df.isnull().sum()/len(df))*100 # Calculo de Porcentaje
nulos = pd.DataFrame({'Valores nulos': valores_nulos, 'Porcentaje': porcentaje_nulos})

#Imprimimos en orden descendente los datos nulos junto a sus porcentajes
nulos.sort_values('Porcentaje', ascending=False)


Valores nulos  Porcentaje
su107c                       8649  100.000000
su107b                       8619   99.653139
padormir2                    8619   99.653139
su116                        8387   96.970748
su107a                       8370   96.774194
padormir1                    8356   96.612325
su112                        8221   95.051451
a313h                        7671   88.692334
a313g                        7671   88.692334
su106                        7597   87.836744
su113                        6262   72.401434
su114                        5814   67.221644
su105                        5419   62.654642
volronquido                  4799   55.486183
cat1                         4799   55.486183
su110                        4799   55.486183
frecron                      4799   55.486183
su109                        4799   55.486183
su111                        4799   55.486183
a1301a                       4496   51.982888
intsel2                      3518   40.675223
rcintura                      583    6.740664
cinturaprom                   552    6.382241
cintura2                      552    6.382241
cintura                       552    6.382241
imc_valid_clas                295    3.410799
imc_valid                     295    3.410799
imc                           281    3.248931
grupoednut                    231    2.670829
a1301                          69    0.797780
a401                           69    0.797780
a502c                          69    0.797780
a502d                          69    0.797780
a502b                          69    0.797780
a502a                          69    0.797780
a301                           69    0.797780
a1312c                         51    0.589664
su101                          10    0.115620
rural                           0    0.000000
tiempo_gen                      0    0.000000
code_upm                        0    0.000000
region_h                        0    0.000000
est_var                         0    0.000000
intsel                          0    0.000000
sexo                            0    0.000000
edad                            0    0.000000
maq                             0    0.000000
su104                           0    0.000000
su103                           0    0.000000
su102                           0    0.000000
su107                           0    0.000000
entidad                         0    0.000000
locali                          0    0.000000
inte                            0    0.000000
idinsert                        0    0.000000
folio                           0    0.000000
su115                           0    0.000000
su108                           0    0.000000
munici                          0    0.000000
horasueno                       0    0.000000
cat1c                           0    0.000000
cat2                            0    0.000000
cat2c                           0    0.000000
hta                             0    0.000000
ponde_f                         0    0.000000
c1                              0    0.000000
calidad                         0    0.000000
sexofem                         0    0.000000
saosc                           0    0.000000
saos                            0    0.000000
insomnio                        0    0.000000
roncamol                        0    0.000000
ronca                           0    0.000000
roncafuerte                     0    0.000000
roncahab                        0    0.000000
cansado                         0    0.000000
cabeceado                       0    0.000000
catberlin2                      0    0.000000
catberlin3                      0    0.000000
catberlin1a                     0    0.000000
apnea                           0    0.000000
catberlin1                      0    0.000000
noreparador                     0    0.000000
berlinriesgo                    0    0.000000
berlinscore                     0    0.000000
catberlin2a                     0    0.000000
berlinscoresinbmi 

Se van a eliminar las columnas que tienen del 50% hacia arriba de valores nulos, por no tener información suficiente para analizar

In [ ]:
umbral = 0.5  # 50% de nulos como criterio
df1 = df.loc[:, df.isnull().mean() < umbral]
df1.head()

folio  inte  idinsert  entidad  munici  locali    maq  edad  \
0  010010001209A0020131     1   1076190        1       1       1  MC036    67   
1  010010001209A0020161     1   1076193        1       1       1  MC036    55   
2  010010001209A0080071     2   1075466        1       1       1  MC034    44   
3  010010001209A0080101     1   1075538        1       1       1  MC034    47   
4  010010001209A0120041     1   1076278        1       1       1  MC036    36   

   sexo  intsel  ...  insom3sem  insomhabi  horasueno1  horasueno2  \
0     1       1  ...          0          0           0           1   
1     1       1  ...          0          0           1           0   
2     0       1  ...          0          0           0           1   
3     0       1  ...          1          1           0           1   
4     1       1  ...          1          1           1           0   

   horasueno3  clonazepam diazepam  alprazolam sedante  cinturaprom  
0           0           1        0           0       1    99.300003  
1           0           0        0           0       0    98.599998  
2           0           0        0           0       0    95.850006  
3           0           0        0           0       0   110.199997  
4           0           0        0           0       0   188.350006  

[5 rows x 79 columns]

Se renombraran 14 variables para facilitar su comprensión

In [ ]:
#Renombramos columnas

df1 = df1.rename(columns={
    'su101': 'hrs_sue_lv',
    'su103': 'calidad_sueno',
    'su107': 'freq_toma_medic',
    'su108': 'precencia_ronquido',
    'su115': 'sleep_driving',
    'a301': 'diabetes',
    'a401': 'hipertension',
    'a502a': 'infarto',
    'a502b': 'angina_pecho',
    'a502c': 'insuficiencia_cardiaca',
    'a502d': 'enfermedad_corazon',
    'a1301': 'fuma',
    'a1312c': 'alcohol',
    'saosc': 'apnea_importante'
})
df1.head()

folio  inte  idinsert  entidad  munici  locali    maq  edad  \
0  010010001209A0020131     1   1076190        1       1       1  MC036    67   
1  010010001209A0020161     1   1076193        1       1       1  MC036    55   
2  010010001209A0080071     2   1075466        1       1       1  MC034    44   
3  010010001209A0080101     1   1075538        1       1       1  MC034    47   
4  010010001209A0120041     1   1076278        1       1       1  MC036    36   

   sexo  intsel  ...  insom3sem  insomhabi  horasueno1  horasueno2  \
0     1       1  ...          0          0           0           1   
1     1       1  ...          0          0           1           0   
2     0       1  ...          0          0           0           1   
3     0       1  ...          1          1           0           1   
4     1       1  ...          1          1           1           0   

   horasueno3  clonazepam diazepam  alprazolam sedante  cinturaprom  
0           0           1        0           0       1    99.300003  
1           0           0        0           0       0    98.599998  
2           0           0        0           0       0    95.850006  
3           0           0        0           0       0   110.199997  
4           0           0        0           0       0   188.350006  

[5 rows x 79 columns]

De la columna 'hrs_sue_lv', se procede a revisar las horas de sueño diarias que duerme una persona al día, cómo se encuentran valores mayores a 24 horas, se procederan a retirar ya que se consideran errores.


In [ ]:
df1['hrs_sue_lv'].value_counts()


hrs_sue_lv
8.0     2855
7.0     1946
6.0     1280
9.0     1108
5.0      544
10.0     469
4.0      169
12.0      82
11.0      82
3.0       59
13.0       9
2.0        8
14.0       6
1.0        4
27.0       2
73.0       2
99.0       2
55.0       1
50.0       1
15.0       1
84.0       1
58.0       1
42.0       1
70.0       1
71.0       1
63.0       1
46.0       1
79.0       1
32.0       1
Name: count, dtype: int64

In [ ]:
# se eliminaran las filas de la columna 'hrs_sue_lv' que tienen valores mayores a 24
df1 = df1[df1['hrs_sue_lv'] <= 24]
df1['hrs_sue_lv'].value_counts()


hrs_sue_lv
8.0     2855
7.0     1946
6.0     1280
9.0     1108
5.0      544
10.0     469
4.0      169
12.0      82
11.0      82
3.0       59
13.0       9
2.0        8
14.0       6
1.0        4
15.0       1
Name: count, dtype: int64

Su101 ahora llamada 'hrs_sue_lv', tiene un 11% de celdas vacias, por lo que se llenaran con la media a través de rangos de edad entre Jovenes (18 a 29 años), Adultos (30 a 59) y Adultos mayores (de los 60 en adelante). Se creara una nueva columna para determinar los rangos de edad ya que se seguiran ocupando

In [ ]:
#creación columna
import numpy as np

condiciones = [
    df1['edad'] <= 29,
    (df1['edad'] >= 30) & (df1['edad'] <= 59),
    df1['edad'] >= 60
]

valores = ['Joven', 'Adulto', 'Adulto Mayor']

df1['rango_edad'] = np.select(condiciones, valores,default='Desconocido')
df1.head()

folio  inte  idinsert  entidad  munici  locali    maq  edad  \
0  010010001209A0020131     1   1076190        1       1       1  MC036    67   
1  010010001209A0020161     1   1076193        1       1       1  MC036    55   
2  010010001209A0080071     2   1075466        1       1       1  MC034    44   
3  010010001209A0080101     1   1075538        1       1       1  MC034    47   
4  010010001209A0120041     1   1076278        1       1       1  MC036    36   

   sexo  intsel  ...  insomhabi  horasueno1  horasueno2  horasueno3  \
0     1       1  ...          0           0           1           0   
1     1       1  ...          0           1           0           0   
2     0       1  ...          0           0           1           0   
3     0       1  ...          1           0           1           0   
4     1       1  ...          1           1           0           0   

   clonazepam  diazepam alprazolam  sedante cinturaprom    rango_edad  
0           1         0          0        1   99.300003  Adulto Mayor  
1           0         0          0        0   98.599998        Adulto  
2           0         0          0        0   95.850006        Adulto  
3           0         0          0        0  110.199997        Adulto  
4           0         0          0        0  188.350006        Adulto  

[5 rows x 80 columns]

In [ ]:
#calculo de la media por rango de edad
media_sueño_por_rango = df1.groupby('rango_edad')['hrs_sue_lv'].mean().round(2)
print(media_sueño_por_rango)

rango_edad
Adulto          7.39
Adulto Mayor    7.49
Joven           7.74
Name: hrs_sue_lv, dtype: float64


In [ ]:
#transformación para rellenar los campos vacios de hr de sueño con la media
df1['hrs_sue_lv'] = df1.groupby('rango_edad')['hrs_sue_lv'].transform(lambda x: x.fillna(x.mean()))


Procederemos a llenar el campo su108 ahora nombrado Ronca con una correlación la variable ronca. Se decidio usarse así ya que ronca es una variable que tiene todos los campos completos con los valores 0 que significa que no ronca y 1 que determina que si ronca, pero la variable su108 a demás de las opciones anteriores, también contiene la opción de "No lo sé", que se me hace una respuesta más real

In [ ]:
print(df1.columns)


Index(['folio', 'inte', 'idinsert', 'entidad', 'munici', 'locali', 'maq',
       'edad', 'sexo', 'intsel', 'intsel2', 'hrs_sue_lv', 'su102',
       'calidad_sueno', 'su104', 'freq_toma_medic', 'precencia_ronquido',
       'sleep_driving', 'tiempo_gen', 'est_var', 'region_h', 'code_upm',
       'rural', 'c1', 'ponde_f', 'diabetes', 'hipertension', 'infarto',
       'angina_pecho', 'insuficiencia_cardiaca', 'enfermedad_corazon', 'fuma',
       'alcohol', 'cintura', 'cintura2', 'rcintura', 'grupoednut', 'imc',
       'imc_valid', 'imc_valid_clas', 'horasueno', 'calidad', 'cat1c', 'cat2',
       'cat2c', 'hta', 'saos', 'apnea_importante', 'sexofem', 'insomnio',
       'ronca', 'roncafuerte', 'roncahab', 'roncamol', 'apnea', 'catberlin1',
       'noreparador', 'cansado', 'cabeceado', 'catberlin2', 'catberlin3',
       'catberlin1a', 'catberlin2a', 'berlinscore', 'berlinriesgo',
       'berlinscoresinbmi', 'berlinriesgosinbmi', 'malacalsueno',
       'impactoinsom', 'insom3sem', 'insomhabi',

In [ ]:
df1.loc[df1['precencia_ronquido'].isna() & (df1['ronca'] == 0), 'precencia_ronquido'] = 1
df1.loc[df1['precencia_ronquido'].isna() & (df1['ronca'] == 1), 'precencia_ronquido'] = 2


Se procedera a eliminar las filas que no contienen valores en los campos de diabetes ya que se encontro que no se cuenta información en las mismas filas (63) de otros padecimientos por lo que no se puede ver el impacto del sueño con las enfermedades

In [ ]:
#visualización de valores nulos
df1['diabetes'].isna().sum()


np.int64(63)

In [ ]:
#eliminamos filas
df1 = df1[df1['diabetes'].notna()]
df1['diabetes'].isna().sum()


np.int64(0)

Para la variable imc_valid_clas que califica si el paciente tiene un peso adecuado, sobrepeso y obesidad, dicha variable tiene el 3.41% de datos nulos, por lo que se procederá a llenar los valores de tres formas diferentes.

La primera opción será correlacionando la variable IMC ya que en el catalogo nos marca que si el IMC es menor a 25 imc_valid_clas es igual a 1, si el IMC es mayor a 25 y menor o igual a 30  es igual a 2 y por último si el IMC es mayor a 30 es iguial a 3.

Pero esto no nos soluciona la cantidad de valores nulos de dicha variable por lo que la segunda opción a tomar es la columna cinturaprom, pude apreciar que esta columna tiene valores disparados entre las clasificaciones imc_valid_clas, por lo que se tomaran las medianas para poder designar el valor máximo a comparar.

Por último para las celdas que sigan vacías, se tomara en cuenta el rango de edad y el sexo para darles una calificación

In [ ]:
#Correlación de acuerdo al IMC
df1.loc[df1['imc_valid_clas'].isna() & (df1['imc'] < 25), 'imc_valid_clas'] = 1
df1.loc[df1['imc_valid_clas'].isna() & (df1['imc'] > 25) & (df1['imc'] <= 30), 'imc_valid_clas'] = 2
df1.loc[df1['imc_valid_clas'].isna() & (df1['imc'] > 30), 'imc_valid_clas'] = 3


In [ ]:
#calculo de la media de la variable cinturaprom
media_cintura = df1.groupby('imc_valid_clas')['cinturaprom'].mean()
print(media_cintura)


imc_valid_clas
1.0     83.034601
2.0     94.202955
3.0    107.183693
Name: cinturaprom, dtype: float64


In [ ]:
mediana_cintura = df.groupby('imc_valid_clas')['cinturaprom'].median()
print(mediana_cintura)

imc_valid_clas
1.0     82.174999
2.0     93.400002
3.0    105.350002
Name: cinturaprom, dtype: float64


In [ ]:
#De acuerdo a las mediciones anteriores, se determino que si cinturaprom es menor o igual a 83 se agregara el valor de 1
#Si es mayor de 83 y menor o igual a 94 se colocará 2
#y mayores de 94 se colocará 3

df1.loc[df1['imc_valid_clas'].isna() & (df1['cinturaprom'] <= 83), 'imc_valid_clas'] = 1
df1.loc[df1['imc_valid_clas'].isna() & (df1['cinturaprom'] > 83) & (df1['cinturaprom'] <= 94), 'imc_valid_clas'] = 2
df1.loc[df1['imc_valid_clas'].isna() & (df1['cinturaprom'] > 94), 'imc_valid_clas'] = 3

In [ ]:
#visualización de valores nulos
df1['imc_valid_clas'].isna().sum()

np.int64(199)

In [ ]:
#calculo de la media de acuerdo al rango de edad y el sexo del paciente
media_imc = df1.groupby(['rango_edad', 'sexo'])['imc'].mean()
print(media_imc)


rango_edad    sexo
Adulto        0       29.870789
              1       28.613298
Adulto Mayor  0       28.819708
              1       26.814286
Joven         0       27.011071
              1       26.526315
Name: imc, dtype: float64


In [ ]:
#Como se puede observar la media en todos los casos esta por arriba de 25 y debajo de 30 por lo que se procederá a calificarlos con 2 al restante de celdas nulas
df1['imc_valid_clas'] = df1['imc_valid_clas'].fillna(2)
df1['imc_valid_clas'].isna().sum()

np.int64(0)

Una vez que ya tenemos completa la información en las columnas deseadas, se
eliminaran 33 variables de columnas que no son relevantes para el análisis de la información por redundancia en la información.

In [ ]:
#Eliminaremos otras columnas que no consideraremos necesarias para nuestro análisis

columnas_a_eliminar = [
    'folio', 'inte', 'idinsert', 'maq', 'intsel', 'intsel2', 'su102', 'su104',
    'tiempo_gen', 'est_var', 'code_upm', 'c1', 'ponde_f', 'cintura', 'cintura2',
    'rcintura', 'grupoednut', 'imc_valid', 'cat1c', 'saos', 'roncafuerte',
    'roncahab', 'roncamol', 'catberlin1', 'noreparador', 'catberlin2',
    'catberlin3', 'catberlin1a', 'catberlin2a', 'berlinscore', 'berlinriesgo',
    'berlinscoresinbmi', 'berlinriesgosinbmi', 'malacalsueno', 'impactoinsom',
    'insomhabi', 'horasueno1', 'horasueno2', 'horasueno3', 'cat2', 'hta', 'sexofem',
    'cat2c','hta','sexofem','ronca','horasueno','calidad','insom3sem'
]

# Get the list of columns that actually exist in df1
existing_columns_to_drop = [col for col in columnas_a_eliminar if col in df1.columns]

df1 = df1.drop(columns=existing_columns_to_drop)
df1.head()

entidad  munici  locali  edad  sexo  hrs_sue_lv  calidad_sueno  \
0        1       1       1    67     1         6.0              2   
1        1       1       1    55     1         8.0              3   
2        1       1       1    44     0         6.0              3   
3        1       1       1    47     0         6.0              3   
4        1       1       1    36     1         8.0              2   

   freq_toma_medic precencia_ronquido  sleep_driving  ...  insomnio  apnea  \
0                2                  1              2  ...         0      0   
1                1                  9              1  ...         0      0   
2                1                  2              2  ...         0      0   
3                1                  1              2  ...         0      0   
4                1                  1              2  ...         0      0   

   cansado  cabeceado  clonazepam  diazepam  alprazolam  sedante  cinturaprom  \
0        0          0           1         0           0        1    99.300003   
1        0          1           0         0           0        0    98.599998   
2        1          0           0         0           0        0    95.850006   
3        1          0           0         0           0        0   110.199997   
4        0          0           0         0           0        0   188.350006   

     rango_edad  
0  Adulto Mayor  
1        Adulto  
2        Adulto  
3        Adulto  
4        Adulto  

[5 rows x 33 columns]

Procederemos a guardar el dataset limpio en un archivo .csv, para posteriormente llevar a cabo la realización del dashboard

In [ ]:
# Guardar el DataFrame como archivo CSV
df1.to_csv('sueño_limpio.csv', index=False)

#Descargar el archivo a tu computadora
from google.colab import files
files.download('sueño_limpio.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
df1.to_excel('sueño_limpio.xlsx', index=False)
from google.colab import files
files.download('sueño_limpio.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>